launch_template/main.tf

```diff
@@ -52,57 +52,77 @@ resource "aws_launch_template" "template" {
   name = "${var.env}-${var.role}"
 
   iam_instance_profile {
-    name = "${var.iam_instance_profile_name}"
+    name = var.iam_instance_profile_name
   }
 
-  image_id = "${lookup(var.ami_id_map, var.role, var.default_ami_id)}"
+  image_id = lookup(var.ami_id_map, var.role, var.default_ami_id)
 
-  instance_initiated_shutdown_behavior = "${var.instance_initiated_shutdown_behavior}"
+  instance_initiated_shutdown_behavior = var.instance_initiated_shutdown_behavior
 
-  instance_type = "${var.instance_type}"
+  instance_type = var.instance_type
 
-  user_data = "${var.user_data}"
+  user_data = var.user_data
 
   monitoring {
     enabled = true
   }
 
-  vpc_security_group_ids = ["${var.security_group_ids}"]
+  vpc_security_group_ids = var.security_group_ids
 
   tag_specifications {
     resource_type = "instance"
-    tags {
-      Name = "asg-${var.env}-${var.role}",
-      prefix = "${var.role}",
+    tags = {
+      Name   = "asg-${var.env}-${var.role}"
+      prefix = var.role
       domain = "${var.env}.${var.root_domain}"
     }
   }
 
   tag_specifications {
     resource_type = "volume"
-    tags {
-      Name = "asg-${var.env}-${var.role}",
-      prefix = "${var.role}",
+    tags = {
+      Name   = "asg-${var.env}-${var.role}"
+      prefix = var.role
       domain = "${var.env}.${var.root_domain}"
     }
   }
 
-  tags = "${
-    merge(
-      map(
-        "prefix", "${var.role}",
-        "domain", "${var.env}.${var.root_domain}"
-      ),
-      var.template_tags
-    )
-  }"
-
-  block_device_mappings = "${var.block_device_mappings}"
+  tags = merge(
+    {
+      "prefix" = var.role
+      "domain" = "${var.env}.${var.root_domain}"
+    },
+    var.template_tags,
+  )
+
+  dynamic "block_device_mappings" {
+    for_each = var.block_device_mappings
+    content {
+      device_name  = lookup(block_device_mappings.value, "device_name", null)
+      no_device    = lookup(block_device_mappings.value, "no_device", null)
+      virtual_name = lookup(block_device_mappings.value, "virtual_name", null)
+
+      dynamic "ebs" {
+        for_each = flatten(list(lookup(block_device_mappings.value, "ebs", [])))
+        content {
+          delete_on_termination = lookup(ebs.value, "delete_on_termination", null)
+          encrypted             = lookup(ebs.value, "encrypted", null)
+          iops                  = lookup(ebs.value, "iops", null)
+          kms_key_id            = lookup(ebs.value, "kms_key_id", null)
+          snapshot_id           = lookup(ebs.value, "snapshot_id", null)
+          volume_size           = lookup(ebs.value, "volume_size", null)
+          volume_type           = lookup(ebs.value, "volume_type", null)
+        }
+      }
+    }
+  }
```

Comment:
Yikes, what a nuisance that we have to do this hack.

It would be good to provide a comment at the top that all of this logic is just to assign `var.block_device_mappings` to the resource's `block_device_mappings`.